bearer-cache.js

```diff
@@ -0,0 +1,121 @@
+const assert = require('assert');
+
+const STATES = {
+  COLD: 'COLD',
+  FETCHING: 'FETCHING',
+  WARM: 'WARM',
+};
+
+/**
+ * The BearerCache implements a simple state machine for managing the lifecycle
+ * of an authentication token.
+ *
+ * When instantiated, a cache will be in the Cold state: it has no token. The
+ * Cold state can transition to the Fetching state, indicating that a request
+ * has been initiated for a token.
+ *
+ * In the Fetching state, callers can queue up for the result by awaiting the
+ * promise returned by #wait. If the fetch fails, the cache can transition to
+ * the Cold state -- anyone awaiting the result will receive a rejection with
+ * the reason. If the fetch succeeds, the cache can transition to the Warm state
+ * -- anyone awaiting the result will receive a resolution with the token.
+ *
+ * In the Warm state, callers can simply query for the token with #value. When
+ * a token is deemed expired, based on a finite lifetime, the cache
+ * automatically transitions back to Cold.
+ *
+ *            ┌───────────┐         ┌───────────┐
+ *            │           ├──Fetch──▶           │
+ * ───Create──▶   Cold    │         │ Fetching  │
+ *            │           ◀──Fail───┤           │
+ *            └─────▲─────┘         └─────┬─────┘
+ *                  │                     │
+ *               Expire  ┌───────────┐ Succeed
+ *                  │    │           │    │
+ *                  └────│   Warm    ◀────┘
+ *                       │           │
+ *                       └───────────┘
+ */
+class BearerCache {
+  constructor() {
+    this.token = null;
+    this.expiresAt = null;
+    this.fetcher = null;
+    this.resolver = null;
+    this.rejecter = null;
+  }
+
+  state() {
+    if (this.fetcher !== null) {
+      return STATES.FETCHING;
+    }
+
+    if (this.token && this.expiresAt > Date.now()) {
```

Comment:
minor: Consider latching this comparison, e.g., set `this.token = null` the first time `this.expiresAt > Date.now()`. 

Aside from the micro-optimization of maybe saving you a syscall (unimportant), it would avoid an unexpected transition from `COLD` to `WARM` if the wall clock changed.

Reply:
It's a good suggestion, I think I've guarded against this with an assert that you can't access `this.token` unless we're in the warm state. To observers using the public api, I think they're identical.